### PR TITLE
agentic rag sk - agents and agentic teams

### PR DESCRIPTION
**Pull request by snova-codym**

### CODE REVIEW

File: utils/search/search_components.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        self.vectorstore = vectorstore
+        self.embeddings = embeddings
+        self.examples: Optional[Dict] = examples
+        
+        self.configs: Dict = self.load_config(configs)
+        self.prompts_paths: Dict = self.configs["prompts"]
+
+    def tavily_web_search(self, state: dict) -> dict:
+        """
+        Web search using Tavily based based on the question
```
Comment (snova-francescar) [inline utils/search/search_components.py:58]: Repetition.

Comment (snova-francescar) [reply utils/search/search_components.py:58]: Period missing.

File: utils/agents/corrective_rag.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    documents : List[str]
+    answers: List[str]
+    original_question: str
+
+    """
+    Represents the state of a RAG (Retreival Augmented Generation) graph.
+
+    Args:
+        question: The question being asked.
+        generation: The generated text.
```
[the remark this answers is not in the document]
Comment (snova-codym) [reply utils/agents/corrective_rag.py:28]: [Added more detail - generated by the llm agent.](generation: The most recent generated text from the LLM agent.)

File: utils/agents/corrective_rag.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        documents: A list of relevant documents.
+        answers: A list of possible answers.
```
[the remark this answers is not in the document]
Comment (snova-codym) [reply utils/agents/corrective_rag.py:30]: answers: A list of possible answers that have been accumulated by the LLM agents.

File: utils/agents/corrective_rag.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                 kwargs: Dict[str,int] = {"recursion_limit": 50}
+                 ) -> tuple[dict[str, Any], dict[str, Any] | Any]:
+        """
+        Calls the RAG (Reasoning and Generation) app to generate an answer to a given question.
+
+        Args:
+            app: The RAG app object.
+            question: The question to be answered.
+            kwargs: Keyword arguments to be passed to the app. 
+            Defaults to {"recursion_limit": 50}
```
[the remark this answers is not in the document]
Comment (snova-codym) [reply utils/agents/corrective_rag.py:130]: Is this an ok style?

def call_rag(self, 
                 app: CompiledStateGraph, 
                 question: str, 
                 kwargs: Dict[str,int] = {"recursion_limit": 50}
                 ) -> tuple[dict[str, Any], dict[str, Any] | Any]:

Note that the indentation is a bit different in VSCode - the args are directly underneath each other.

File: utils/agents/corrective_rag.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                - "source_documents" (List[str]): A list of source documents used 
+                to generate the answer.
+        """
+        
+        runnable = RunnableConfig(configurable={"question": question})
+
+        response = {}
+        output = app.invoke(runnable, kwargs=kwargs)
+        response["answer"] = output["generation"]
+        sources = [o for o in output["documents"]]
```
[the remark this answers is not in the document]
Comment (snova-codym) [reply utils/agents/corrective_rag.py:146]: Good point!  Changing for simplicity.

File: utils/agents/return_message.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        self.configs: Dict = self.load_config(configs)
+        self.prompts_paths: Dict = self.configs['prompts']
+
+    def init_return_message(self) -> None:
+        """
+        Initializes the return message by loading the prompt
+        and combining it with the LLM and a string
+        output parser.
+
+        Args:
```
[the remark this answers is not in the document]
Comment (snova-codym) [reply utils/agents/return_message.py:62]: Sorry - are you asking for reference or to also remove Args and Returns?

File: utils/agents/return_message.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            dict: The state dictionary containing updated the response to the user.
+        """
+
+        question: str = state['question']
+        answer: str = state['generation']
+        next: str = state['next']
+
+        response: str = self.return_message.invoke({'question': question, 'answer': answer, 'next': next})
+        print('---RESPONSE---')
+        print(response)
```
[the remark this answers is not in the document]
Comment (snova-codym) [reply utils/agents/return_message.py:89]: Waiting on a future update so I don't introduce too many changes just yet.  Should be replaced with logging, though!

File: utils/agents/corrective_rag.py [hunk 8]
```diff
@@ -116,38 +101,34 @@ def initialize(self) -> None:
...
         Calls the RAG (Reasoning and Generation) app to generate an answer to a given question.
 
         Args:
             app: The RAG app object.
             question: The question to be answered.
-            kwargs: Keyword arguments to be passed to the app. 
-            Defaults to {"recursion_limit": 50}  
-            Recursion limit controls how many runnables to invoke without 
+            kwargs: Keyword arguments to be passed to the app.
+            Defaults to {"recursion_limit": 50}
```
Comment (snova-francescar) [inline utils/agents/corrective_rag.py:115]: Period.

File: utils/agents/supervisor.py [hunk 9]
```diff
@@ -57,33 +55,28 @@ def init_supervisor_router(self) -> None:
...
 
-
     def supervisor_router(self, state: dict) -> dict:
         """
         This method is the supervisor router, which handles the state of the conversation.
         Args:
-        state: A dictionary containing the current state
-        of the conversation, including the question, question history,
-        answer history, and teams.
+            state: A dictionary containing the current state
```
Comment (snova-francescar) [inline utils/agents/supervisor.py:67]: Do not indent the first line.

File: utils/agents/supervisor.py [hunk 10]
```diff
@@ -57,33 +55,28 @@ def init_supervisor_router(self) -> None:
...
+            of the conversation, including the question, question history,
+            answer history, and teams.
         Returns:
-        The state dictionary with the updated next step in the conversation,
-        as determined by the supervisor.
+            The state dictionary with the updated next step in the conversation,
```
Comment (snova-francescar) [inline utils/agents/supervisor.py:71]: Idem.

File: utils/agents/teams/corrective_rag_team.py [hunk 11]
```diff
@@ -19,21 +19,31 @@
 
 
 class SuperState(TypedDict):
+    """
+    A dictionary representing the current state of the supervisor.
+
+    Args:
+        question: The user question being asked.
```
Comment (snova-francescar) [inline utils/agents/teams/corrective_rag_team.py:26]: Idem.

Comment (snova-francescar) [inline utils/agents/teams/corrective_rag_team.py:26]: No need to indent here.

File: utils/agents/teams/return_message_team.py [hunk 12]
```diff
@@ -18,6 +17,15 @@
 
 
 class ReturnTeamState(TypedDict):
+    """
+    A typed dictionary representing the state of a team in a conversation.
+
+    Args:
+        question: The question being asked by the user.
```
Comment (snova-francescar) [inline utils/agents/teams/return_message_team.py:24]: Idem.

File: utils/agents/teams/tavily_search_team.py [hunk 13]
```diff
@@ -19,6 +19,17 @@
 
 
 class SearchGraphState(TypedDict):
+    """
+    Represents the state of a search graph.
+
+    Args:
+        question: The original question being asked by the user.
```
Comment (snova-francescar) [inline utils/agents/teams/tavily_search_team.py:26]: Idem.

File: utils/search/search_components.py [hunk 14]
```diff
@@ -27,11 +27,12 @@ class SearchComponents(BaseComponents):
...
 
-    def __init__(self, 
-                configs: str, 
-                embeddings: object, 
-                vectorstore: Chroma, 
-                examples: dict=None) -> None:
+    def __init__(
+        self, 
+        configs: str, 
+        embeddings: object,
```
Comment (snova-francescar) [inline utils/search/search_components.py:33]: Object instead of Embeddings?

**Pull request merged**